datacube/model/__init__.py

```diff
@@ -840,3 +927,183 @@ def metadata_from_doc(doc: Mapping[str, Any]) -> MetadataType:
     from .fields import get_dataset_fields
     MetadataType.validate(doc)  # type: ignore
     return MetadataType(doc, get_dataset_fields(doc))
+
+
+class ExtraDimensions:
+    """
+    Definition for the additional dimensions between (t) and (y, x)
+
+    It allows the creation of a subsetted ExtraDimensions that contains slicing information relative to
+    the original dimension coordinates.
+    """
+
+    def __init__(self, extra_dim: Dict[str, Any]):
+        """Init function
+
+        :param extra_dim: Dimension definition dict, typically retrieved from the product definition's
+            `extra_dimensions` field.
+        """
+        import xarray
+
+        # Dict of information about each dimension
+        self._dims = extra_dim
+        # Dimension slices that results in this ExtraDimensions object
+        self._dim_slice = {
+            name: (0, len(dim['values'])) for name, dim in extra_dim.items()
+        }
+        # Coordinate information
+        self._coords = {
+            name: xarray.DataArray(
+                data=dim['values'],
+                coords={name: dim['values']},
+                dims=(name,),
+                name=name,
+            ).astype(dim['dtype'])
+            for name, dim in extra_dim.items()
+        }
+
+    def has_empty_dim(self) -> bool:
+        """Return True if ExtraDimensions has an empty dimension, otherwise False.
+
+        :return: A boolean if ExtraDimensions has an empty dimension, otherwise False.
+        """
+        for value in self._coords.values():
+            if value.shape[0] == 0:
+                return True
+        return False
+
+    def __getitem__(self, dim_slices: Dict[str, Union[float, Tuple[float, float]]]) -> "ExtraDimensions":
+        """Return a ExtraDimensions subsetted by dim_slices
+
+        :param dim_slices: Optional cache to re-use geoboxes instead of creating new one each time
+        :return: An ExtraDimensions object subsetted by `dim_slices`
+        """
+        # Check all dimensions specified in dim_slices exists
+        unknown_keys = set(dim_slices.keys()) - set(self._dims.keys())
+        if unknown_keys:
+            raise KeyError(f"Found unknown keys {unknown_keys} in dim_slices")
+
+        from copy import deepcopy
+
+        ed = ExtraDimensions(deepcopy(self._dims))
+        ed._dim_slice = self._dim_slice
+
+        # Convert to integer index
+        for dim_name, dim_slice in dim_slices.items():
+            dim_slices[dim_name] = self.coord_slice(dim_name, dim_slice)
+
+        for dim_name, dim_slice in dim_slices.items():
+            # Adjust slices relative to original.
+            if dim_name in ed._dim_slice:
+                ed._dim_slice[dim_name] = (
+                    ed._dim_slice[dim_name][0] + dim_slice[0],
+                    ed._dim_slice[dim_name][0] + dim_slice[1],
```

Comment:
Not sure what this does, but are we sure it's not meant to be this:

```suggestion
                    ed._dim_slice[dim_name][0] + dim_slice[0],
                    ed._dim_slice[dim_name][1] + dim_slice[1],
```

Reply:
This should be correct, since there's only 1 dimension, there's only 1 origin / point of reference.

Reply:
Done, Please let me know if there are any other required changes.